### PR TITLE
Run jars without process/exec

### LINKED ISSUE #46: No need to run uberjar via babashka.process/exec
`bbin install <uberjar>` will create a wrapper script that relies on `babashka.process/exec` to execute the jar. This means that babashka is invoked twice, **adding extra startup time for the script** (also see #45). 

After some personal testing and a [conversation with @borkdude in Slack](https://clojurians.slack.com/archives/C0400TZENE7/p1672951967367479) I believe this is unnecessary.

### Proposed alternative
- [The wrapper script](https://github.com/babashka/bbin/blob/main/src/babashka/bbin/scripts.clj#L135) would need to be adapted to something like:

```clojure
(require '[babashka.classpath :refer [add-classpath]])
(add-classpath "the-uber-jar.jar")
(require '[the-main-ns])
(apply the-main-ns/-main *command-line-args*)
```

- The install process has to find the main class, and fail if it cannot find the main-ns (and maybe also when it cannot find a -main in that namespace).

I've played around with the compiled code from Babashka itself to find the main ns and I found that this runs in interpreted Babashka as well.  ([here](https://github.com/babashka/babashka/blob/bc0b59500bdb58f710ae1c684167cdb188963218/src/babashka/impl/classpath.clj#L82) and [here](https://github.com/babashka/babashka/blob/bc0b59500bdb58f710ae1c684167cdb188963218/src/babashka/impl/classpath.clj#L31-L42))

I'll create a PR for this issue somewhere the coming days unless there are reasons for `babashka.process/exec` that I missed.

### PR DESCRIPTION
**Pull request by jeroenvandijk**
Resolves #46 

Test with `bb gen-script && bb bbin install <your-jar>`

I'm leaving this as a draft until I get to the following
- [x] Double check if Babashka exposes [java.util.jar.Attributes/getValue](https://docs.oracle.com/javase/7/docs/api/java/util/jar/Attributes.html#getValue(java.util.jar.Attributes.Name)) and if not create an issue ([created here](https://github.com/babashka/babashka/issues/1463))
- [X] Check if tests are running (`bb ci`)
- [ ] Maybe add new tests

Please answer the following questions and leave the below in as part of your PR.

- [X] This PR corresponds to an [issue with a clear problem statement](https://github.com/babashka/babashka/blob/master/doc/dev.md#start-with-an-issue-before-writing-code).
  - https://github.com/babashka/bbin/issues/46

### CODE REVIEW

**Comment by jeroenvandijk** [comment]
I've marked it "ready for review" and I'm expecting some suggestions for improvement. Especially for the testing part I'm curious what you think.

Also the not sure what to do with the update of `bbin` file

File: src/babashka/bbin/scripts.clj [hunk 1]
```diff
@@ -292,32 +293,58 @@
...
+
+
+(defn jar->main-ns [jar-path]
+  (let [jar-file (io/file jar-path)
+        manifest-path "META-INF/MANIFEST.MF"]
+    (or (with-open [jar (java.util.jar.JarFile. jar-file)]
+          (when (.getEntry jar manifest-path)
+            (main-ns (java.net.URL. "jar" nil (str "file:" (.getAbsolutePath jar-file) "!/" manifest-path)))))
+
+        (throw (ex-info "jar has no Main-Class" {:babashka/exit 1})))))
```
Comment (rads) [inline src/babashka/bbin/scripts.clj:315]: The current implementation is causing some [test failures in Windows](https://github.com/babashka/bbin/actions/runs/3891827078/jobs/6721062643). I was able to fix this by simplifying to the following code:
```clojure
(defn jar->main-ns [jar-path]
  (with-open [jar-file (JarFile. (fs/file jar-path))]
    (or (some-> (.getManifest jar-file)
                (.getMainAttributes)
                ;; FIXME .getValue is missing in Babashka? Create issue.
                ;; See https://docs.oracle.com/javase/7/docs/api/java/util/jar/Attributes.html
                #_(.getValue "Main-Class")
                (->> (into {}) (some (fn [[k v]]
                                       (when (clojure.string/includes? k "Main-Class") v))))
                (clojure.main/demunge))
        (throw (ex-info "jar has no Main-Class" {:babashka/exit 1})))))
```

Comment (borkdude) [reply src/babashka/bbin/scripts.clj:315]: Note that the Attributes class was merged to master, but hasn't been released yet. This workaround seems good though.

**Review by rads — CHANGES_REQUESTED** [review]
- Update `jar->main-ns` to fix errors in Windows
- Merge `main` to bring branch up-to-date

File: src/babashka/bbin/scripts.clj (hunk 1 above)
Comment (rads) [reply src/babashka/bbin/scripts.clj:315]: @borkdude: Would you prefer to leave in the workaround for now or bump the minimum Babashka version for bbin?

Comment (borkdude) [reply src/babashka/bbin/scripts.clj:315]: Workaround + note and then remove workaround in 6 months or so

Comment (rads) [reply src/babashka/bbin/scripts.clj:315]: @jeroenvandijk: Let's update the comment here based on @borkdude's feedback. ^

**Comment by rads** [comment]
@jeroenvandijk: Thanks for the PR!

> I've marked it "ready for review" and I'm expecting some suggestions for improvement. Especially for the testing part I'm curious what you think.

The change looks good overall. We have tests for `bbin install *.jar` so that's enough for me. I try not to add any more tests than necessary to keep maintenance down to a minimum. That said, is there anything you think we should add?

> Also the not sure what to do with the update of bbin file

Use the `bb dev` task to start a watcher to update the `bbin` file in the background. It's best to ensure this is in sync within each commit. I usually just leave `bb dev` running until I'm done working on `bbin`.

**Comment by jeroenvandijk** [comment]
@rads Thank you for the feedback and simplification suggestion. I've applied the feedback.

I have also added an extra step to the install of a remote jar. After this change if the remote jar doesn't have a Main-Class the jar doesn't get cached in `.babashka/bbin/jars`. E.g. you can try `bbin install https://repo1.maven.org/maven2/org/clojure/clojure/1.11.1/clojure-1.11.1.jar`

I couldn't run all the tests locally due to the dependency of a private repo:
```
{:args ("git" "clone" "--quiet" "--mirror" "git@github.com:rads/bbin-test-lib-private.git" "/Users/jeroen/.gitlibs/_repos/ssh/github.com/rads/bbin-test-lib-private"), :exit 128, :out "", :err "ERROR: Repository not found.\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n"}
 at sci.lang.Var.invoke (lang.cljc:182)
    sci.impl.analyzer$return_call$reify__4815.eval (analyzer.cljc:1351)
    sci.impl.analyzer$analyze_throw$reify__4571.eval (analyzer.cljc:937)
    sci.impl.analyzer$return_if$reify__4526.eval (analyzer.cljc:819)
```

Regarding the extra tests, maybe testing with a jar without a "Main-Class" and confirm bbin will fail? For http and local jars.

**Comment by rads** [comment]
The existing tests are fixed now. I'm going to merge this in and do the final test/clean up as follow-up commits so I can finish this tonight.

**Review by rads — APPROVED** [review]

**Comment by jeroenvandijk** [comment]
Thanks @rads! I see you had to make some adjustments and even a fix for re-installing a jar. Should have caught that myself!    Thanks again for `bbin`.

**Comment by rads** [comment]
All good, thanks for the contribution!
